specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+  - [L2CrossDomainMessenger Proxy Update](#l2crossdomainmessenger-proxy-update)
+  - [L2StandardBridge Proxy Update](#l2standardbridge-proxy-update)
+- [Selection of `ETHER_TOKEN_ADDRESS`](#selection-of-ether_token_address)
+- [Security Considerations](#security-considerations)
+  - [OptimismPortal Token Allowance](#optimismportal-token-allowance)
+  - [Decimal Scaling](#decimal-scaling)
+  - [Interoperability Support](#interoperability-support)
+  - [Wrapped Ether](#wrapped-ether)
+- [Future Directions](#future-directions)
+  - [Enabling a L2 Native ERC20 Token to Pay for Gas](#enabling-a-l2-native-erc20-token-to-pay-for-gas)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+Custom gas token is also known as "custom L2 native asset". It allows for an ERC20 token to be collateralized on L1
+to back the native asset on L2. The native asset is used to buy gas which is used to pay for resources on the network
+and is represented by EVM opcodes such as `CALLVALUE` (`msg.value`).
+
+Both the L1 and L2 smart contract systems MUST be able to introspect on the address of the gas paying token to
+ensure the security of the system. The source of truth for the configuration of the address of the token is
+in the L1 `SystemConfig` smart contract.
+
+## Constants
+
+| Constant          | Value                           | Description |
+| ----------------- | ------------------------------- | ----------- |
+| `ETHER_TOKEN_ADDRESS` |  `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` | Represents ether for gas paying asset |
+| `DEPOSITOR_ACCOUNT` | `address(0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001)` | Account with auth permissions on `L1Block` contract |
+| `GAS_PAYING_TOKEN_SLOT` | `bytes32(uint256(keccak256("opstack.gaspayingtoken")) - 1)` | Storage slot that contains the address and decimals of the gas paying token |
+
+## Properties of a Gas Paying Token
+
+The gas paying token MUST satisfy the standard [ERC20](https://eips.ethereum.org/EIPS/eip-20) interface
+and implementation. It MUST be a L1 contract.
+
+The gas paying token cannot:
+
+- Have a fee on transfer
+- Have rebasing logic
+- Have call hooks on transfer
+- Have more than 18 decimals
+- Have out of band methods for modifying balance or allowance
+
+It MUST be enforced on chain that the token has 18 or less decimals to guarantee no losses of precision when
+depositing a token that has a different number of decimals.
+
+## Configuring the Gas Paying Token
+
+The gas paying token is set within the L1 `SystemConfig` smart contract. This allows for easy access to the information
+required to know if an OP Stack chain is configured to use a custom gas paying token. The gas paying token is set
+during initialization and cannot be modified until another initialization. It is assumed that the chain operator
+will not modify the gas paying token address unless they specifically decide to do so and appropriately handle collateralization.
```

Comment:
This decision stems from a lack of standardization on the `initialize` logic. Right now we generally reset everything on an `initialize`. Ideally we move away from the `initialize` pattern in the future because of this and have "versioned storage" with migrations between the versions, where the logic for the storage migrations live in their own contract

Reply:
I think it is generally more simple to just set the contract address again in `initialize` which will trigger the setter call on the L1Block contract. Removing that ability adds an extra test case (initialize again, see no change) but it does remove a footgun where it is changed accidentally. Our upgrade tooling should be able to assert that there is no state diff for that slot. Using a 2 step process can still allow for the change in token - just set to `address(0)` using `StorageSetter.setBytes32` and then pass in a new token address into the initializer

Reply:
I am fine to update the spec + implementation with this given my thoughts above, although we will have this sort of problem as long as we use `initialize` - how do we maintain consistency and sanity with what slots are set and not set?

Reply:
> Using a 2 step process can still allow for the change in token - just set to `address(0)` using `StorageSetter.setBytes32` and then pass in a new token address into the initializer

True, this is a great counterargument. Backing to up think of my original motivation here, in general the situation is:
- For chains that don't want to be standard chains, they can change their implementations and do whatever they want, so logic preventing this does not matter.
- For chains that do want to be standard chains, the implementation is fixed and logic preventing this helps them avoid mistakes.

So perhaps a potential solution in the initialize method is (pseudocode that ignores `GAS_PAYING_TOKEN_SLOT` for simplicity):
```solidity
function initialize(address _gasPayingToken) external {
  // `gasPayingToken` is the value in storage.
  require(gasPayingToken == address(0) || gasPayingToken == _gasPayingToken);
  gasPayingToken = _gasPayingToken;
}
```

This way, even if chains want to be standard chains but do their upgrades without using our upgrade tooling, they still get guarantees the token address won't change. They can work around this with StorageSetter like you said, but it seems extremely unlikely to *accidentally* do that, so I think something like this would be sufficient.

> set the contract address again in `initialize` which will trigger the setter call on the L1Block contract.

For my own understanding, what's the mechanism through which this works, i.e. how specifically does setting the storage slot to the pre-existing value trigger the L1Block setter call?

Reply:
We have tried a similar pattern in the past and it adds complexity with downgrades or reinit to the same implementation. We ended up bricking the contract in a way where the source code needed to be modified before a reinitialize was possible 😅 

I am definitely open to following this approach as its semantics are clear but then we will need to be mindful about the semantics forever, which I don't think is that bad in this case. Adding in a gas optimization to not sstore if its the same is also something i would like to avoid

Reply:
I noted the requirement of the above pattern directly in the spec